### ISSUE TITLE
Version Control between Knora Salsah, Sipi and GraphDB

### ISSUE THREAD

**Issue by janCstoffregen** [dhlab-basel/Knora#728]
Dear Reader,
I don't have a clue which version of the software mentioned above works with one another. 

I get in trouble when I update only one part of the software mentioned above, then the whole set of software might not work together anymore. For example, I just tried to update Salsah, but it does not work anymore with a version of Knora being two months old. In addition, sometimes all current develop branches don't work with one another. (eg sometimes salsah only works with a wip branch)

If I update Knora, I might have to update the ontology, since the Knora Base ontology changes once in a while as well. It ends up that I try to wait with my updates as long as possible although I would like to perform updates much more often.

I wrote a script that installs all pieces of software within only a couple of lines:
https://github.com/nie-ine/Scripts/tree/master/docker-compose-deployment-Knora-Salsah-Sipi-GraphDB
It uses only existing docker images on docker hub, so I know that everything works fine as a set.

I wonder what you think about the following:
- Create a new git repository only for the installation process, with only one docker-compose file that will include all parts of the software. @kilchenmann , the docker file (not for production yet though) is here: https://github.com/nie-ine/Scripts/blob/master/docker-compose-deployment-Knora-Salsah-Sipi-GraphDB/Salsah/Dockerfile
- Automatically push docker images on docker hub by each software with travis and perform integration testing between all parts of the software

If you create the repository I will gladly implement this.

Thank you very much and best regards,
Jan

**Comment by loicjaouen** [comment]
super cool compose script!

I would recommend to use [tagged releases](https://github.com/dhlab-basel/Knora/releases) instead of [commits](https://github.com/nie-ine/Scripts/tree/master/docker-compose-deployment-Knora-Salsah-Sipi-GraphDB/Knora) .

And the Knora/Salsah/Sipi dev teams should be able to tag releases and say which ones work with which ones.

**Comment by benjamingeer** [comment]
All the develop branches are supposed to work together. If they don't, that's a bug, and we would be glad if you opened an issue to let us know.

We don't yet have official releases, but when we do, we're supposed to write release notes saying what's changed and what you have to do to upgrade.

**Comment by janCstoffregen** [comment]
@loicjaouen 
that would be amazing! If we would have tagged docker images on docker hub for each piece software and a simple table which ones work together, that would be amazing.

@benjamingeer 
The problem is, that a version of Knora, being 2 months old, does not work with the current Salsah anymore. I don't want to update Knora though bc that will take time since the ontology might have changed and I don't have time to investigate this. Thus, I wonder which version of Salsah works with my older version of Knora. 
To sum up, to successfully update one piece of the software, you have to update all parts of the software (and test if all branches work together). This is sometimes very very time intensive.
Updating only one part of the software leads to the question which newer version I can use since the latest does not work. This question can be solved only be trial and error (with newer but not latest versions) which is time intensive as well.

**Comment by benjamingeer** [comment]
> a version of Knora, being 2 months old, does not work with the current Salsah anymore

Unless you report a specific bug, there's no way I can help you with this.

**Comment by subotic** [comment]
This all is already on our (mine) todo list :-)

The docker images published on docker hub are generated on Travis after all tests (also integration tests) run successfully. At the moment, these tests don't include [Salsah2](https://github.com/dhlab-basel/Salsah).

The idea that I have, is to add `Salsah2` integration testing to the Knora repository. This way, we can make sure that all `develop` branches of Knora, Salsa, and Sipi will work together.

Additionally, I would like to add integration tests to the Salsah repository, to make sure that Salsah `develop` works with Knora `develop`. The current `develop` version of Salsah contains a set of service integration tests for the Admin API. But as long as there are no other tests in Salsah, we cannot automate this process.

At the moment, it can happen that Knora and Salsah `develop` don't work with each other, since we don't have an automated way of testing it.

In the past, I have tagged some releases of Knora, Salsah, and Sipi that worked together, but nothing recently.

As for which version works with which version, can be difficult at the moment, as we don't have so many released versions.

**Comment by subotic** [comment]
One way would be to maintain a list of commit hashes in each repository denoting with which version it is compatible at the moment. But this would be a pain to do.

**Comment by janCstoffregen** [comment]
Can we do this once per month maybe? As you say, it is painful to do so but it is also painful if each user has to find it out individually as well.

**Comment by janCstoffregen** [comment]
Could we open one deployment - repository for all parts where those kinds of bugs can be registered? Otherwise I never know in which repository I can report those things as well since I don't know which software causes the error (sometimes knora reports an error but sipi is the problem)

**Comment by benjamingeer** [comment]
> I never know in which repository I can report those things as well since I don't know which software causes the error

Just make your best guess, and we'll figure it out.

**Comment by subotic** [comment]
> Can we do this once per month maybe? As you say, it is painful to do so but it is also painful if each user has to find it out individually as well.

The plan was to do it at least once a month. At the moment it takes a bit longer since I'm adding a few changes that require existing data to be changed, and wanted to bundle these changes into the same release.

**Comment by janCstoffregen** [comment]
@subotic ok, great, thank you very much for this!! I can track the releases and test it with the other parts respectively and will try to post it to the right repository if I get a bug.

Please let me know if I can help in any way regarding the docker-compose deployment.

**Comment by janCstoffregen** [comment]
Dear dhlab,
I am proud to announce that we got further with the development and can start knora, salsah, sipi, graphdb and nie-os just with one command now (docker-compose up).

https://github.com/nie-ine/Quick-Installation-Guide---Knora-Salsah-Sipi-GraphDB/tree/master/PrebuiltImages/20180619

A proxy is part of the docker-compose as well which eases the deployment. The ontologies are already stored persistently so that the user does not have to import those him- or herself.

This will be our solution for the version control for now, since I test all parts of the software together before releasing a new version. Can I close this issue for now?

**Comment by subotic** [comment]
great, I will close the issue then.
